Cannot create a snowflake_user with provider 0.70.1

**Provider Version**

0.70.1
terraform-provider-snowflake_v0.70.1

**Terraform Version**

Terraform v1.1.9

**Describe the bug**

Creating a new snowflake user resource fails on apply with a plugin did not respond error.

Example plan:
```
# module.account.snowflake_user.user will be created
  + resource "snowflake_user" "user" {
      + default_role       = "ROLE"
      + default_warehouse  = "WAREHOUSE"
      + disabled           = false
      + display_name       = (sensitive value)
      + email              = (sensitive value)
      + first_name         = (sensitive value)
      + has_rsa_public_key = (known after apply)
      + id                 = (known after apply)
      + last_name          = (sensitive value)
      + login_name         = (known after apply)
      + name               = (sensitive value)
    }

Plan: 1 to add, 0 to change, 0 to destroy.
```

After applying
```
Error: Plugin did not respond
│
│   with module.account.snowflake_user.user,
│   on ../../../modules/account/users.tf line 1344, in resource "snowflake_user" "user":
│ 1344: resource "snowflake_user" "user" {
│
│ The plugin encountered an error, and failed to respond to the
│ plugin.(*GRPCProvider).ApplyResourceChange call. The plugin logs may
│ contain more details.

Stack trace from the terraform-provider-snowflake_v0.70.1 plugin:

panic: interface conversion: sdk.ObjectIdentifier is sdk.DatabaseObjectIdentifier, not sdk.AccountObjectIdentifier

goroutine 85 [running]:
github.com/Snowflake-Labs/terraform-provider-snowflake/pkg/resources.ReadUser(0xc00050d600, {0x1ac67c0?, 0xc000b40000?})
	github.com/Snowflake-Labs/terraform-provider-snowflake/pkg/resources/user.go:244 +0x96a
github.com/Snowflake-Labs/terraform-provider-snowflake/pkg/resources.CreateUser(0x0?, {0x1ac67c0?, 0xc000b40000?})
	github.com/Snowflake-Labs/terraform-provider-snowflake/pkg/resources/user.go:236 +0xbc8
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).create(0x1de2900?, {0x1de2900?, 0xc000e7b7d0?}, 0xd?, {0x1ac67c0?, 0xc000b40000?})
	github.com/hashicorp/terraform-plugin-sdk/v2@v2.26.1/helper/schema/resource.go:695 +0x178
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).Apply(0xc00025cc40, {0x1de2900, 0xc000e7b7d0}, 0xc00087cdd0, 0xc00050d480, {0x1ac67c0, 0xc000b40000})
	github.com/hashicorp/terraform-plugin-sdk/v2@v2.26.1/helper/schema/resource.go:837 +0xa85
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*GRPCProviderServer).ApplyResourceChange(0xc000277920, {0x1de2900?, 0xc000e7b6b0?}, 0xc00019a640)
	github.com/hashicorp/terraform-plugin-sdk/v2@v2.26.1/helper/schema/grpc_provider.go:1021 +0xe8d
github.com/hashicorp/terraform-plugin-go/tfprotov5/tf5server.(*server).ApplyResourceChange(0xc00085c640, {0x1de2900?, 0xc000e7acc0?}, 0xc000aabe30)
	github.com/hashicorp/terraform-plugin-go@v0.14.3/tfprotov5/tf5server/server.go:818 +0x574
github.com/hashicorp/terraform-plugin-go/tfprotov5/internal/tfplugin5._Provider_ApplyResourceChange_Handler({0x1a79800?, 0xc00085c640}, {0x1de2900, 0xc000e7acc0}, 0xc000aabdc0, 0x0)
	github.com/hashicorp/terraform-plugin-go@v0.14.3/tfprotov5/internal/tfplugin5/tfplugin5_grpc.pb.go:385 +0x170
google.golang.org/grpc.(*Server).processUnaryRPC(0xc000768960, {0x1df22c0, 0xc0008e0000}, 0xc000526fc0, 0xc0005afe30, 0x2a3c3e0, 0x0)
	google.golang.org/grpc@v1.53.0/server.go:1336 +0xd33
google.golang.org/grpc.(*Server).handleStream(0xc000768960, {0x1df22c0, 0xc0008e0000}, 0xc000526fc0, 0x0)
	google.golang.org/grpc@v1.53.0/server.go:1704 +0xa36
google.golang.org/grpc.(*Server).serveStreams.func1.2()
	google.golang.org/grpc@v1.53.0/server.go:965 +0x98
created by google.golang.org/grpc.(*Server).serveStreams.func1
	google.golang.org/grpc@v1.53.0/server.go:963 +0x28a

Error: The terraform-provider-snowflake_v0.70.1 plugin crashed!

This is always indicative of a bug within the plugin. It would be immensely
helpful if you could report the crash with the plugin's maintainers so that it
can be fixed. The output above should help diagnose the issue.

Operation failed: failed running terraform apply (exit 1)
```

**Expected behavior**

Expected user to be created in Snowflake without an error - sometimes the user is created sometimes it is not.

**Code samples and commands**

Example TF code:
```
resource "snowflake_user" "user" {
  name     = "USER"
  email    = "USER@EMAIL.COM"
  disabled = false

  first_name   = "Test"
  last_name    = "User"
  display_name = "Test User"

  default_warehouse = "WAREHOUSE"
  default_role      = "ROLE"
}
```

I'm getting a similar error as well, with 0.74
Although the plugin crashed, _some_ user resources were provisioned
Rolling back to 0.69 allowed the apply to succeed, though it subsequently failed when two of the four users existed already (I manually imported them to state)

`Stack trace from the terraform-provider-snowflake_v0.74.0 plugin:
panic: interface conversion: sdk.ObjectIdentifier is sdk.TableColumnIdentifier, not sdk.AccountObjectIdentifier
goroutine 112 [running]:
github.com/Snowflake-Labs/terraform-provider-snowflake/pkg/resources.ReadUser(0xc0005b1980, {0x1c66fc0?, 0xc0005b6a90?})
	github.com/Snowflake-Labs/terraform-provider-snowflake/pkg/resources/user.go:244 +0x96a
github.com/Snowflake-Labs/terraform-provider-snowflake/pkg/resources.CreateUser(0x0?, {0x1c66fc0?, 0xc0005b6a90?})
	github.com/Snowflake-Labs/terraform-provider-snowflake/pkg/resources/user.go:236 +0xbc8
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).create(0x1fb9f80?, {0x1fb9f80?, 0xc0007dd2c0?}, 0xd?, {0x1c66fc0?, 0xc0005b6a90?})
	github.com/hashicorp/terraform-plugin-sdk/v2@v2.29.0/helper/schema/resource.go:766 +0x178
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).Apply(0xc0001f5880, {0x1fb9f80, 0xc0007dd2c0}, 0xc0005b7380, 0xc0005b1800, {0x1c66fc0, 0xc0005b6a90})
	github.com/hashicorp/terraform-plugin-sdk/v2@v2.29.0/helper/schema/resource.go:909 +0xa7e
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*GRPCProviderServer).ApplyResourceChange(0xc000e90018, {0x1fb9f80?, 0xc0007dd1a0?}, 0xc0007f2410)
	github.com/hashicorp/terraform-plugin-sdk/v2@v2.29.0/helper/schema/grpc_provider.go:1060 +0xe8d
github.com/hashicorp/terraform-plugin-go/tfprotov5/tf5server.(*server).ApplyResourceChange(0xc000e9a000, {0x1fb9f80?, 0xc0007dc990?}, 0xc0005b8a80)
	github.com/hashicorp/terraform-plugin-go@v0.19.0/tfprotov5/tf5server/server.go:859 +0x574
github.com/hashicorp/terraform-plugin-go/tfprotov5/internal/tfplugin5._Provider_ApplyResourceChange_Handler({0x1c16380?, 0xc000e9a000}, {0x1fb9f80, 0xc0007dc990}, 0xc0005b8a10, 0x0)
	github.com/hashicorp/terraform-plugin-go@v0.19.0/tfprotov5/internal/tfplugin5/tfplugin5_grpc.pb.go:467 +0x170
google.golang.org/grpc.(*Server).processUnaryRPC(0xc0000314a0, {0x1fc9c00, 0xc000e84b60}, 0xc0007ab9e0, 0xc000e881b0, 0x2cc4558, 0x0)
	google.golang.org/grpc@v1.58.2/server.go:1376 +0xdd2
google.golang.org/grpc.(*Server).handleStream(0xc0000314a0, {0x1fc9c00, 0xc000e84b60}, 0xc0007ab9e0, 0x0)
	google.golang.org/grpc@v1.58.2/server.go:1753 +0xa36
google.golang.org/grpc.(*Server).serveStreams.func1.1()
	google.golang.org/grpc@v1.58.2/server.go:998 +0x98
created by google.golang.org/grpc.(*Server).serveStreams.func1
	google.golang.org/grpc@v1.58.2/server.go:996 +0x18c
Error: The terraform-provider-snowflake_v0.74.0 plugin crashed!`

I'm having the same issue with the provider from version 0.70.1 up to 0.73.0:

```
Stack trace from the terraform-provider-snowflake_v0.73.0 plugin:

panic: interface conversion: sdk.ObjectIdentifier is sdk.SchemaObjectIdentifier, not sdk.AccountObjectIdentifier

goroutine 40 [running]:
github.com/Snowflake-Labs/terraform-provider-snowflake/pkg/resources.ReadUser(0xc00064b980, {0x1b6e6a0?, 0xc000882680?})
github.com/Snowflake-Labs/terraform-provider-snowflake/pkg/resources/user.go:244 +0x96a
github.com/Snowflake-Labs/terraform-provider-snowflake/pkg/resources.CreateUser(0x0?, {0x1b6e6a0?, 0xc000882680?})
github.com/Snowflake-Labs/terraform-provider-snowflake/pkg/resources/user.go:236 +0xbc8
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).create(0x1ea1900?, {0x1ea1900?, 0xc000ca0090?}, 0xd?, {0x1b6e6a0?, 0xc000882680?})
github.com/hashicorp/terraform-plugin-sdk/v2@v2.29.0/helper/schema/resource.go:766 +0x178
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).Apply(0xc000739ea0, {0x1ea1900, 0xc000ca0090}, 0xc000752f70, 0xc00064b800, {0x1b6e6a0, 0xc000882680})
github.com/hashicorp/terraform-plugin-sdk/v2@v2.29.0/helper/schema/resource.go:909 +0xa7e
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*GRPCProviderServer).ApplyResourceChange(0xc000871dd0, {0x1ea1900?, 0xc000c87f50?}, 0xc000674c80)
github.com/hashicorp/terraform-plugin-sdk/v2@v2.29.0/helper/schema/grpc_provider.go:1060 +0xe8d
github.com/hashicorp/terraform-plugin-go/tfprotov5/tf5server.(*server).ApplyResourceChange(0xc0000f7b80, {0x1ea1900?, 0xc000c87560?}, 0xc00068b340)
github.com/hashicorp/terraform-plugin-go@v0.19.0/tfprotov5/tf5server/server.go:859 +0x574
github.com/hashicorp/terraform-plugin-go/tfprotov5/internal/tfplugin5._Provider_ApplyResourceChange_Handler({0x1b1fda0?, 0xc0000f7b80}, {0x1ea1900, 0xc000c87560}, 0xc00068b2d0, 0x0)
github.com/hashicorp/terraform-plugin-go@v0.19.0/tfprotov5/internal/tfplugin5/tfplugin5_grpc.pb.go:467 +0x170
google.golang.org/grpc.(*Server).processUnaryRPC(0xc000696d20, {0x1eb1480, 0xc00073e1a0}, 0xc000c8f320, 0xc000643050, 0x2b61258, 0x0)
google.golang.org/grpc@v1.57.0/server.go:1360 +0xe23
google.golang.org/grpc.(*Server).handleStream(0xc000696d20, {0x1eb1480, 0xc00073e1a0}, 0xc000c8f320, 0x0)
google.golang.org/grpc@v1.57.0/server.go:1737 +0xa36
google.golang.org/grpc.(*Server).serveStreams.func1.1()
google.golang.org/grpc@v1.57.0/server.go:982 +0x98
created by google.golang.org/grpc.(*Server).serveStreams.func1
google.golang.org/grpc@v1.57.0/server.go:980 +0x18c

Error: The terraform-provider-snowflake_v0.73.0 plugin crashed!

This is always indicative of a bug within the plugin. It would be immensely
helpful if you could report the crash with the plugin's maintainers so that it
can be fixed. The output above should help diagnose the issue.
```

--- a/pkg/helpers/helpers.go
+++ b/pkg/helpers/helpers.go
@@ -65,15 +65,26 @@ func EncodeSnowflakeID(attributes ...interface{}) string {
 	// is attribute already an object identifier?
 	if len(attributes) == 1 {
 		if id, ok := attributes[0].(sdk.ObjectIdentifier); ok {
+			if val := reflect.ValueOf(id); val.Kind() == reflect.Ptr && val.IsNil() {
+				log.Panicf("Nil object identifier received")
+			}
 			parts := make([]string, 0)
 			switch v := id.(type) {
 			case sdk.AccountObjectIdentifier:
 				parts = append(parts, v.Name())
+			case *sdk.AccountObjectIdentifier:
+				parts = append(parts, v.Name())
 			case sdk.DatabaseObjectIdentifier:
+				parts = append(parts, v.DatabaseName(), v.Name())
+			case *sdk.DatabaseObjectIdentifier:
 				parts = append(parts, v.DatabaseName(), v.Name())
 			case sdk.SchemaObjectIdentifier:
 				parts = append(parts, v.DatabaseName(), v.SchemaName(), v.Name())
+			case *sdk.SchemaObjectIdentifier:
+				parts = append(parts, v.DatabaseName(), v.SchemaName(), v.Name())
 			case sdk.TableColumnIdentifier:
+				parts = append(parts, v.DatabaseName(), v.SchemaName(), v.TableName(), v.Name())
+			case *sdk.TableColumnIdentifier:
 				parts = append(parts, v.DatabaseName(), v.SchemaName(), v.TableName(), v.Name())
 			default:
 				log.Panicf("Unsupported object identifier: %v", id)

--- a/pkg/helpers/helpers.go
+++ b/pkg/helpers/helpers.go
@@ -66,15 +66,15 @@ func EncodeSnowflakeID(attributes ...interface{}) string {
 	if len(attributes) == 1 {
 		if id, ok := attributes[0].(sdk.ObjectIdentifier); ok {
 			parts := make([]string, 0)
-			switch id.(type) {
+			switch v := id.(type) {
 			case sdk.AccountObjectIdentifier:
-				parts = append(parts, id.(sdk.AccountObjectIdentifier).Name())
+				parts = append(parts, v.Name())
 			case sdk.DatabaseObjectIdentifier:
-				parts = append(parts, id.(sdk.DatabaseObjectIdentifier).DatabaseName(), id.(sdk.DatabaseObjectIdentifier).Name())
+				parts = append(parts, v.DatabaseName(), v.Name())
 			case sdk.SchemaObjectIdentifier:
-				parts = append(parts, id.(sdk.SchemaObjectIdentifier).DatabaseName(), id.(sdk.SchemaObjectIdentifier).SchemaName(), id.(sdk.SchemaObjectIdentifier).Name())
+				parts = append(parts, v.DatabaseName(), v.SchemaName(), v.Name())
 			case sdk.TableColumnIdentifier:
-				parts = append(parts, id.(sdk.TableColumnIdentifier).DatabaseName(), id.(sdk.TableColumnIdentifier).SchemaName(), id.(sdk.TableColumnIdentifier).TableName(), id.(sdk.TableColumnIdentifier).Name())
+				parts = append(parts, v.DatabaseName(), v.SchemaName(), v.TableName(), v.Name())
 			default:
 				log.Panicf("Unsupported object identifier: %v", id)
 			}

--- a/pkg/helpers/helpers.go
+++ b/pkg/helpers/helpers.go
@@ -65,10 +65,18 @@ func EncodeSnowflakeID(attributes ...interface{}) string {
 	// is attribute already an object identifier?
 	if len(attributes) == 1 {
 		if id, ok := attributes[0].(sdk.ObjectIdentifier); ok {
-			// remove quotes and replace dots with pipes
-			parts := strings.Split(id.FullyQualifiedName(), ".")
-			for i, part := range parts {
-				parts[i] = strings.Trim(part, `"`)
+			parts := make([]string, 0)
+			switch id.(type) {
+			case sdk.AccountObjectIdentifier:
+				parts = append(parts, id.(sdk.AccountObjectIdentifier).Name())
+			case sdk.DatabaseObjectIdentifier:
+				parts = append(parts, id.(sdk.DatabaseObjectIdentifier).DatabaseName(), id.(sdk.DatabaseObjectIdentifier).Name())
+			case sdk.SchemaObjectIdentifier:
+				parts = append(parts, id.(sdk.SchemaObjectIdentifier).DatabaseName(), id.(sdk.SchemaObjectIdentifier).SchemaName(), id.(sdk.SchemaObjectIdentifier).Name())
+			case sdk.TableColumnIdentifier:
+				parts = append(parts, id.(sdk.TableColumnIdentifier).DatabaseName(), id.(sdk.TableColumnIdentifier).SchemaName(), id.(sdk.TableColumnIdentifier).TableName(), id.(sdk.TableColumnIdentifier).Name())
+			default:
+				log.Panicf("Unsupported object identifier: %v", id)
 			}
 			return strings.Join(parts, IDDelimiter)
 		}

--- a/pkg/helpers/helpers_test.go
+++ b/pkg/helpers/helpers_test.go
@@ -1,6 +1,7 @@
 package helpers
 
 import (
+	"fmt"
 	"testing"
 
 	"github.com/Snowflake-Labs/terraform-provider-snowflake/pkg/sdk"
@@ -76,7 +77,7 @@ func TestDecodeSnowflakeParameterID(t *testing.T) {
 // TODO: add tests for non object identifiers
 func TestEncodeSnowflakeID(t *testing.T) {
 	testCases := map[string]struct {
-		identifier        sdk.ObjectIdentifier
+		identifier        any
 		expectedEncodedID string
 	}{
 		"encodes account object identifier": {
@@ -91,6 +92,10 @@ func TestEncodeSnowflakeID(t *testing.T) {
 			identifier:        sdk.NewAccountObjectIdentifier("data.base"),
 			expectedEncodedID: `data.base`,
 		},
+		"encodes pointer to account object identifier": {
+			identifier:        sdk.Pointer(sdk.NewAccountObjectIdentifier("database")),
+			expectedEncodedID: `database`,
+		},
 		"encodes database object identifier": {
 			identifier:        sdk.NewDatabaseObjectIdentifier("database", "schema"),
 			expectedEncodedID: `database|schema`,
@@ -102,6 +107,10 @@ func TestEncodeSnowflakeID(t *testing.T) {
 		"encodes database object identifier with dots": {
 			identifier:        sdk.NewDatabaseObjectIdentifier("data.base", "sche.ma"),
 			expectedEncodedID: `data.base|sche.ma`,
+		},
+		"encodes pointer to database object identifier": {
+			identifier:        sdk.Pointer(sdk.NewDatabaseObjectIdentifier("database", "schema")),
+			expectedEncodedID: `database|schema`,
 		},
 		"encodes schema object identifier": {
 			identifier:        sdk.NewSchemaObjectIdentifier("database", "schema", "table"),
@@ -115,6 +124,10 @@ func TestEncodeSnowflakeID(t *testing.T) {
 			identifier:        sdk.NewSchemaObjectIdentifier("data.base", "sche.ma", "tab.le"),
 			expectedEncodedID: `data.base|sche.ma|tab.le`,
 		},
+		"encodes pointer to schema object identifier": {
+			identifier:        sdk.Pointer(sdk.NewSchemaObjectIdentifier("database", "schema", "table")),
+			expectedEncodedID: `database|schema|table`,
+		},
 		"encodes table column identifier": {
 			identifier:        sdk.NewTableColumnIdentifier("database", "schema", "table", "column"),
 			expectedEncodedID: `database|schema|table|column`,
@@ -127,6 +140,10 @@ func TestEncodeSnowflakeID(t *testing.T) {
 			identifier:        sdk.NewTableColumnIdentifier("data.base", "sche.ma", "tab.le", "col.umn"),
 			expectedEncodedID: `data.base|sche.ma|tab.le|col.umn`,
 		},
+		"encodes pointer to table column identifier": {
+			identifier:        sdk.Pointer(sdk.NewTableColumnIdentifier("database", "schema", "table", "column")),
+			expectedEncodedID: `database|schema|table|column`,
+		},
 	}
 
 	for name, tc := range testCases {
@@ -137,10 +154,26 @@ func TestEncodeSnowflakeID(t *testing.T) {
 	}
 
 	t.Run("panics for unsupported object identifier", func(t *testing.T) {
-		require.Panics(t, func() {
-			EncodeSnowflakeID(unsupportedObjectIdentifier{})
+		id := unsupportedObjectIdentifier{}
+		require.PanicsWithValue(t, fmt.Sprintf("Unsupported object identifier: %v", id), func() {
+			EncodeSnowflakeID(id)
 		})
 	})
+
+	nilTestCases := []any{
+		(*sdk.AccountObjectIdentifier)(nil),
+		(*sdk.DatabaseObjectIdentifier)(nil),
+		(*sdk.SchemaObjectIdentifier)(nil),
+		(*sdk.TableColumnIdentifier)(nil),
+	}
+
+	for i, tt := range nilTestCases {
+		t.Run(fmt.Sprintf("handle nil pointer to object identifier %d", i), func(t *testing.T) {
+			require.PanicsWithValue(t, "Nil object identifier received", func() {
+				EncodeSnowflakeID(tt)
+			})
+		})
+	}
 }
 
 type unsupportedObjectIdentifier struct{}

--- a/pkg/helpers/helpers_test.go
+++ b/pkg/helpers/helpers_test.go
@@ -77,7 +77,7 @@ func TestDecodeSnowflakeParameterID(t *testing.T) {
 // TODO: add tests for non object identifiers
 func TestEncodeSnowflakeID(t *testing.T) {
 	testCases := map[string]struct {
-		identifier        any
+		identifier        sdk.ObjectIdentifier
 		expectedEncodedID string
 	}{
 		"encodes account object identifier": {

--- a/pkg/helpers/helpers_test.go
+++ b/pkg/helpers/helpers_test.go
@@ -80,40 +80,40 @@ func TestEncodeSnowflakeID(t *testing.T) {
 		expectedEncodedID string
 	}{
 		"encodes account object identifier": {
-			identifier:        sdk.NewAccountObjectIdentifier("account"),
-			expectedEncodedID: `account`,
+			identifier:        sdk.NewAccountObjectIdentifier("database"),
+			expectedEncodedID: `database`,
 		},
 		"encodes quoted account object identifier": {
-			identifier:        sdk.NewAccountObjectIdentifier("\"account\""),
-			expectedEncodedID: `account`,
+			identifier:        sdk.NewAccountObjectIdentifier("\"database\""),
+			expectedEncodedID: `database`,
 		},
 		"encodes account object identifier with a dot": {
-			identifier:        sdk.NewAccountObjectIdentifier("acc.ount"),
-			expectedEncodedID: `acc.ount`,
+			identifier:        sdk.NewAccountObjectIdentifier("data.base"),
+			expectedEncodedID: `data.base`,
 		},
 		"encodes database object identifier": {
-			identifier:        sdk.NewDatabaseObjectIdentifier("account", "database"),
-			expectedEncodedID: `account|database`,
+			identifier:        sdk.NewDatabaseObjectIdentifier("database", "schema"),
+			expectedEncodedID: `database|schema`,
 		},
 		"encodes quoted database object identifier": {
-			identifier:        sdk.NewDatabaseObjectIdentifier("\"account\"", "\"database\""),
-			expectedEncodedID: `account|database`,
+			identifier:        sdk.NewDatabaseObjectIdentifier("\"database\"", "\"schema\""),
+			expectedEncodedID: `database|schema`,
 		},
 		"encodes database object identifier with dots": {
-			identifier:        sdk.NewDatabaseObjectIdentifier("acc.ount", "data.base"),
-			expectedEncodedID: `acc.ount|data.base`,
+			identifier:        sdk.NewDatabaseObjectIdentifier("data.base", "sche.ma"),
+			expectedEncodedID: `data.base|sche.ma`,
 		},
 		"encodes schema object identifier": {
-			identifier:        sdk.NewSchemaObjectIdentifier("account", "database", "schema"),
-			expectedEncodedID: `account|database|schema`,
+			identifier:        sdk.NewSchemaObjectIdentifier("database", "schema", "table"),
+			expectedEncodedID: `database|schema|table`,
 		},
 		"encodes quoted schema object identifier": {
-			identifier:        sdk.NewSchemaObjectIdentifier("\"account\"", "\"database\"", "\"schema\""),
-			expectedEncodedID: `account|database|schema`,
+			identifier:        sdk.NewSchemaObjectIdentifier("\"database\"", "\"schema\"", "\"table\""),
+			expectedEncodedID: `database|schema|table`,
 		},
 		"encodes schema object identifier with dots": {
-			identifier:        sdk.NewSchemaObjectIdentifier("acc.ount", "data.base", "sche.ma"),
-			expectedEncodedID: `acc.ount|data.base|sche.ma`,
+			identifier:        sdk.NewSchemaObjectIdentifier("data.base", "sche.ma", "tab.le"),
+			expectedEncodedID: `data.base|sche.ma|tab.le`,
 		},
 		"encodes table column identifier": {
 			identifier:        sdk.NewTableColumnIdentifier("database", "schema", "table", "column"),
@@ -135,4 +135,20 @@ func TestEncodeSnowflakeID(t *testing.T) {
 			require.Equal(t, tc.expectedEncodedID, encodedID)
 		})
 	}
+
+	t.Run("panics for unsupported object identifier", func(t *testing.T) {
+		require.Panics(t, func() {
+			EncodeSnowflakeID(unsupportedObjectIdentifier{})
+		})
+	})
+}
+
+type unsupportedObjectIdentifier struct{}
+
+func (i unsupportedObjectIdentifier) Name() string {
+	return "name"
+}
+
+func (i unsupportedObjectIdentifier) FullyQualifiedName() string {
+	return "fully qualified name"
 }

--- a/pkg/helpers/helpers_test.go
+++ b/pkg/helpers/helpers_test.go
@@ -3,6 +3,7 @@ package helpers
 import (
 	"testing"
 
+	"github.com/Snowflake-Labs/terraform-provider-snowflake/pkg/sdk"
 	"github.com/stretchr/testify/require"
 )
 
@@ -70,4 +71,68 @@ func TestDecodeSnowflakeParameterID(t *testing.T) {
 		_, err := DecodeSnowflakeParameterID(id)
 		require.Errorf(t, err, "incompatible identifier: %s", id)
 	})
+}
+
+// TODO: add tests for non object identifiers
+func TestEncodeSnowflakeID(t *testing.T) {
+	testCases := map[string]struct {
+		identifier        sdk.ObjectIdentifier
+		expectedEncodedID string
+	}{
+		"encodes account object identifier": {
+			identifier:        sdk.NewAccountObjectIdentifier("account"),
+			expectedEncodedID: `account`,
+		},
+		"encodes quoted account object identifier": {
+			identifier:        sdk.NewAccountObjectIdentifier("\"account\""),
+			expectedEncodedID: `account`,
+		},
+		"encodes account object identifier with a dot": {
+			identifier:        sdk.NewAccountObjectIdentifier("acc.ount"),
+			expectedEncodedID: `acc.ount`,
+		},
+		"encodes database object identifier": {
+			identifier:        sdk.NewDatabaseObjectIdentifier("account", "database"),
+			expectedEncodedID: `account|database`,
+		},
+		"encodes quoted database object identifier": {
+			identifier:        sdk.NewDatabaseObjectIdentifier("\"account\"", "\"database\""),
+			expectedEncodedID: `account|database`,
+		},
+		"encodes database object identifier with dots": {
+			identifier:        sdk.NewDatabaseObjectIdentifier("acc.ount", "data.base"),
+			expectedEncodedID: `acc.ount|data.base`,
+		},
+		"encodes schema object identifier": {
+			identifier:        sdk.NewSchemaObjectIdentifier("account", "database", "schema"),
+			expectedEncodedID: `account|database|schema`,
+		},
+		"encodes quoted schema object identifier": {
+			identifier:        sdk.NewSchemaObjectIdentifier("\"account\"", "\"database\"", "\"schema\""),
+			expectedEncodedID: `account|database|schema`,
+		},
+		"encodes schema object identifier with dots": {
+			identifier:        sdk.NewSchemaObjectIdentifier("acc.ount", "data.base", "sche.ma"),
+			expectedEncodedID: `acc.ount|data.base|sche.ma`,
+		},
+		"encodes table column identifier": {
+			identifier:        sdk.NewTableColumnIdentifier("database", "schema", "table", "column"),
+			expectedEncodedID: `database|schema|table|column`,
+		},
+		"encodes quoted table column identifier": {
+			identifier:        sdk.NewTableColumnIdentifier("\"database\"", "\"schema\"", "\"table\"", "\"column\""),
+			expectedEncodedID: `database|schema|table|column`,
+		},
+		"encodes table column identifier with dots": {
+			identifier:        sdk.NewTableColumnIdentifier("data.base", "sche.ma", "tab.le", "col.umn"),
+			expectedEncodedID: `data.base|sche.ma|tab.le|col.umn`,
+		},
+	}
+
+	for name, tc := range testCases {
+		t.Run(name, func(t *testing.T) {
+			encodedID := EncodeSnowflakeID(tc.identifier)
+			require.Equal(t, tc.expectedEncodedID, encodedID)
+		})
+	}
 }

--- a/pkg/resources/testdata/TestAcc_ExternalTable_basic/test.tf
+++ b/pkg/resources/testdata/TestAcc_ExternalTable_basic/test.tf
@@ -1,22 +1,22 @@
 resource "snowflake_storage_integration" "i" {
-  name = var.name
+  name                      = var.name
   storage_allowed_locations = [var.location]
-  storage_provider = "S3"
-  storage_aws_role_arn = var.aws_arn
+  storage_provider          = "S3"
+  storage_aws_role_arn      = var.aws_arn
 }
 
 resource "snowflake_stage" "test" {
-  name = var.name
-  url = var.location
-  database = var.database
-  schema = var.schema
+  name                = var.name
+  url                 = var.location
+  database            = var.database
+  schema              = var.schema
   storage_integration = snowflake_storage_integration.i.name
 }
 
 resource "snowflake_external_table" "test_table" {
-  name = var.name
+  name     = var.name
   database = var.database
-  schema = var.schema
+  schema   = var.schema
   comment  = "Terraform acceptance test"
   column {
     name = "column1"
@@ -29,5 +29,5 @@ resource "snowflake_external_table" "test_table" {
     as   = "($1:\"CreatedDate\"::timestamp)"
   }
   file_format = "TYPE = CSV"
-  location = "@\"${var.database}\".\"${var.schema}\".\"${snowflake_stage.test.name}\""
+  location    = "@\"${var.database}\".\"${var.schema}\".\"${snowflake_stage.test.name}\""
 }

--- a/pkg/resources/user_acceptance_test.go
+++ b/pkg/resources/user_acceptance_test.go
@@ -162,3 +162,26 @@ resource "snowflake_user" "w" {
 	log.Printf("[DEBUG] s2 %s", s)
 	return fmt.Sprintf(s, prefix, prefix)
 }
+
+func TestAcc_User_withDotInID(t *testing.T) {
+	r := require.New(t)
+	prefix := "tst-terraform" + strings.ToUpper(acctest.RandStringFromCharSet(10, acctest.CharSetAlpha)) + "user.123"
+	sshkey1, err := testhelpers.Fixture("userkey1")
+	r.NoError(err)
+	sshkey2, err := testhelpers.Fixture("userkey2")
+	r.NoError(err)
+
+	resource.Test(t, resource.TestCase{
+		Providers:    acc.TestAccProviders(),
+		PreCheck:     func() { acc.TestAccPreCheck(t) },
+		CheckDestroy: nil,
+		Steps: []resource.TestStep{
+			{
+				Config: uConfig(prefix, sshkey1, sshkey2),
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckResourceAttr("snowflake_user.w", "name", prefix),
+				),
+			},
+		},
+	})
+}

--- a/pkg/resources/user_acceptance_test.go
+++ b/pkg/resources/user_acceptance_test.go
@@ -163,7 +163,10 @@ resource "snowflake_user" "w" {
 	return fmt.Sprintf(s, prefix, prefix)
 }
 
-func TestAcc_User_withDotInID(t *testing.T) {
+// TestAcc_User_issue2058 proves https://github.com/Snowflake-Labs/terraform-provider-snowflake/issues/2058 issue.
+// The problem was with a dot in user identifier.
+// Before the fix it results in panic: interface conversion: sdk.ObjectIdentifier is sdk.DatabaseObjectIdentifier, not sdk.AccountObjectIdentifier error.
+func TestAcc_User_issue2058(t *testing.T) {
 	r := require.New(t)
 	prefix := "tst-terraform" + strings.ToUpper(acctest.RandStringFromCharSet(10, acctest.CharSetAlpha)) + "user.123"
 	sshkey1, err := testhelpers.Fixture("userkey1")

--- a/pkg/sdk/identifier_helpers.go
+++ b/pkg/sdk/identifier_helpers.go
@@ -121,7 +121,9 @@ type AccountObjectIdentifier struct {
 }
 
 func NewAccountObjectIdentifier(name string) AccountObjectIdentifier {
-	return AccountObjectIdentifier{name: name}
+	return AccountObjectIdentifier{
+		name: strings.Trim(name, `"`),
+	}
 }
 
 func NewAccountObjectIdentifierFromFullyQualifiedName(fullyQualifiedName string) AccountObjectIdentifier {
@@ -268,7 +270,12 @@ type TableColumnIdentifier struct {
 }
 
 func NewTableColumnIdentifier(databaseName, schemaName, tableName, columnName string) TableColumnIdentifier {
-	return TableColumnIdentifier{databaseName: databaseName, schemaName: schemaName, tableName: tableName, columnName: columnName}
+	return TableColumnIdentifier{
+		databaseName: strings.Trim(databaseName, `"`),
+		schemaName:   strings.Trim(schemaName, `"`),
+		tableName:    strings.Trim(tableName, `"`),
+		columnName:   strings.Trim(columnName, `"`),
+	}
 }
 
 func NewTableColumnIdentifierFromFullyQualifiedName(fullyQualifiedName string) TableColumnIdentifier {

--- a/pkg/sdk/privileges.go
+++ b/pkg/sdk/privileges.go
@@ -186,12 +186,12 @@ const (
 
 	// -- For ICEBERG TABLE
 	SchemaObjectPrivilegeApplyBudget SchemaObjectPrivilege = "APPLYBUDGET"
-	//SchemaObjectPrivilegeDelete      SchemaObjectPrivilege = "DELETE" (duplicate)
-	//SchemaObjectPrivilegeInsert      SchemaObjectPrivilege = "INSERT" (duplicate)
-	//SchemaObjectPrivilegeReferences  SchemaObjectPrivilege = "REFERENCES" (duplicate)
-	//SchemaObjectPrivilegeSelect      SchemaObjectPrivilege = "SELECT" (duplicate)
-	//SchemaObjectPrivilegeTruncate      SchemaObjectPrivilege = "Truncate" (duplicate)
-	//SchemaObjectPrivilegeUpdate      SchemaObjectPrivilege = "Update" (duplicate)
+	// SchemaObjectPrivilegeDelete      SchemaObjectPrivilege = "DELETE" (duplicate)
+	// SchemaObjectPrivilegeInsert      SchemaObjectPrivilege = "INSERT" (duplicate)
+	// SchemaObjectPrivilegeReferences  SchemaObjectPrivilege = "REFERENCES" (duplicate)
+	// SchemaObjectPrivilegeSelect      SchemaObjectPrivilege = "SELECT" (duplicate)
+	// SchemaObjectPrivilegeTruncate      SchemaObjectPrivilege = "Truncate" (duplicate)
+	// SchemaObjectPrivilegeUpdate      SchemaObjectPrivilege = "Update" (duplicate)
 
 	// -- For PIPE
 	// { MONITOR | OPERATE } [ , ... ]

